Prop and Erased

tsao-chi opened the pull request:
This pr
+ Added ErasedTerm
+ generate ErasedTerm on inherit/synthesize whenever sort is Prop
+ implement illegal erased term check.
+ add a reduction rule: `elim(erased)` become `erased`
+ corrected prop-related tyck rules for sigma types and struct
+ updated LittleTyper to support new universe system for `type.computeType()` use cases

ice1000 commented on base/src/main/java/org/aya/core/term/ElimTerm.java:
```diff
@@ -27,6 +32,13 @@ public sealed interface ElimTerm extends Term {
         return new CallTerm.Hole(hole.ref(), hole.ulift(), hole.contextArgs(), hole.args().appended(app.arg()));
     }
+    if (app.of() instanceof ErasedTerm erased) {
+      if (erased.type() instanceof FormTerm.Pi pi) {
```
Probably replace with assertion... This should always hold

tsao-chi replied:
Maybe erased.type() could be an ErrorTerm

ice1000 replied:
Oh

ice1000 replied:
Can you add this to the comment?

tsao-chi replied:
ok

ice1000 commented on base/src/main/java/org/aya/core/term/ElimTerm.java:
```diff
@@ -62,6 +74,13 @@ record Proj(@NotNull Term of, int ix) implements ElimTerm {
       return tup.items().get(proj.ix - 1);
     }
+    if (proj.of instanceof ErasedTerm erased) {
+      if (erased.type() instanceof FormTerm.Sigma sigma) {
```
Ditto

ice1000 commented on base/src/main/java/org/aya/tyck/ExprTycker.java:
```diff
@@ -467,6 +466,20 @@ private TermResult checkBoundaries(Expr expr, FormTerm.Path path, Subst subst, T
+    } else if (x.kind() == SortKind.Type || y.kind() == SortKind.Type) {
+      return new FormTerm.Type(lift);
+    } else if (x instanceof FormTerm.ISet && y instanceof FormTerm.ISet) {
+      return FormTerm.ISet.INSTANCE; // TODO: should we allow this?
```
I believe so.

tsao-chi replied:
ok

ice1000 commented on base/src/main/java/org/aya/tyck/ExprTycker.java:
```diff
@@ -467,6 +466,20 @@ private TermResult checkBoundaries(Expr expr, FormTerm.Path path, Subst subst, T
     });
   }
 
+  private FormTerm.Sort calculateSigma(@NotNull FormTerm.Sort x, @NotNull FormTerm.Sort y) {
```
max

tsao-chi replied:
It is not max anymore. It calculates possible sort for a sigma, which is sometimes not the max

ice1000 replied:
Does it also work for pi?

tsao-chi replied:
no

ice1000 replied:
Why

tsao-chi replied:
There is sortPi() for pi

tsao-chi replied:
, which is adapted from agda

ice1000 replied:
Cool

ice1000 commented on base/src/main/java/org/aya/core/term/ElimTerm.java:
```diff
@@ -10,13 +10,31 @@
  */
 public sealed interface ElimTerm extends Term {
+  // ErasedTerm itself and ElimTerm with `of` erased are erased.
+  static @Nullable ErasedTerm checkErased(@NotNull Term term) {
```
```suggestion
  static @Nullable ErasedTerm underlyingErased(@NotNull Term term) {
```

ice1000 commented on base/src/main/java/org/aya/core/term/ElimTerm.java:
```diff
@@ -10,13 +10,31 @@
+  }
+  // ErasedTerm with a Prop type might safely appear in IntroTerms.
+  // ErasedTerm with a non-Prop type or ElimTerm with `of` erased are disallowed.
+  static @Nullable ErasedTerm checkErasedNotProp(@NotNull Term term) {
+    if (term instanceof ElimTerm elim) return checkErased(elim.of());
+    if (term instanceof CallTerm.Access elim) return checkErased(elim.of());
+    return term instanceof ErasedTerm erased && !erased.isProp() ? erased : null;
```
Are you sure? I guess it might be

```suggestion
  static @Nullable ErasedTerm checkErasedNotProp(@NotNull Term term) {
    if (term instanceof ElimTerm elim) return checkErasedNotProp(elim.of());
    if (term instanceof CallTerm.Access elim) return checkErasedNotProp(elim.of());
    return term instanceof ErasedTerm erased && !erased.isProp() ? erased : null;
```

tsao-chi replied:
Yes
Consider the following case
```
A : Prop
a : A
a.v : Nat
```
only things like `f a` are allowed

ice1000 reviewed (commented):
Trying to understand

ice1000 commented on base/src/main/java/org/aya/core/visitor/BetaExpander.java:
```diff
@@ -38,6 +38,11 @@ public interface BetaExpander extends EndoFunctor {
+        if (app.of() instanceof ErasedTerm) {
+          var xi = app.cube().params();
+          var ui = app.args().map(Arg::term);
+          yield new ErasedTerm(app.cube().type().subst(new Subst(xi, ui)));
```
Are you indicating that we're going to use XTT?

ice1000 reviewed (commented):
Trying to understand

tsao-chi commented:
This pr
+ Added ErasedTerm
+ implement illegal erased term check.
+ add a reduction rule: `elim(erased)` become `erased`
+ corrected prop-related tyck rules for sigma types and struct
+ updated LittleTyper to support new universe system for `type.computeType()` use cases

ice1000 commented on base/src/main/java/org/aya/tyck/StmtTycker.java:
```diff
@@ -73,7 +73,7 @@ private void tracing(@NotNull Consumer<Trace.@NotNull Builder> consumer) {
         yield decl.body.fold(
           body -> {
-            var nobody = tycker.inherit(body, signature.result()).wellTyped();
+            var nobody = tycker.checkIllegalErasure(body.sourcePos(), tycker.inherit(body, signature.result())).wellTyped();
```
What about inserting this check in `inherit`?

ice1000 commented:
TODOs:

+ [ ] Erased pathlam

ice1000 commented:
bors r+

ice1000 commented:
birs r+

ice1000 commented:
bors r+
